synapse/handlers/federation.py

```diff
@@ -549,15 +550,29 @@ async def do_invite_join(
                 #   https://github.com/matrix-org/synapse/issues/12998
                 await self.store.store_partial_state_room(room_id, ret.servers_in_room)
 
-            max_stream_id = await self._federation_event_handler.process_remote_join(
-                origin,
-                room_id,
-                auth_chain,
-                state,
-                event,
-                room_version_obj,
-                partial_state=ret.partial_state,
-            )
+            try:
+                max_stream_id = (
+                    await self._federation_event_handler.process_remote_join(
+                        origin,
+                        room_id,
+                        auth_chain,
+                        state,
+                        event,
+                        room_version_obj,
+                        partial_state=ret.partial_state,
+                    )
+                )
+            except PartialStateConflictError as e:
+                # The homeserver was already in the room and it is no longer partial
+                # stated. We ought to be doing a local join instead. Turn the error into
+                # a 503, as a hint to the client to try again.
+                # TODO(faster_joins): `_should_perform_remote_join` suggests that we may
+                #   do a remote join for restricted rooms even if we have full state.
+                logger.error(
+                    "Room %s was un-partial stated while processing remote join.",
+                    room_id,
+                )
+                raise SynapseError(HTTPStatus.SERVICE_UNAVAILABLE, e.msg, e.errcode)
```

Comment:
We should never his this path right now, but might in future?

My reading of this method (`do_invite_join`) is that it assumes that the homeserver is not in the room,
while `_should_perform_remote_join` seems to imply that we might end up here even if the homeserver is already in the room?

Reply:
(I can't see where the `_should_perform_remote_join` that you're referring to is)

Reply:
It's used a few levels up in the call stack, inside `update_membership_locked`:
https://github.com/matrix-org/synapse/blob/755261524728c4a8c3f48c91590bdaef0731651e/synapse/handlers/room_member.py#L937
https://github.com/matrix-org/synapse/blob/755261524728c4a8c3f48c91590bdaef0731651e/synapse/handlers/room_member.py#L817-L843

Reply:
Ah, restricted rooms means that this can be hit even though we're already in the room. I guess I see (and I also see why I would never have known about it, because they're a recent addition and I don't really know how they work).

I think the code here is reasonable anyway.